File: examples/aps2_demo.m
function aps2_demo(addr)

aps = APS2();
connect(aps, addr);

[ver, ver_str, git_sha1, build_timestamp] = get_firmware_version(aps);

fprintf('Connected to APS2 running firmware version %s and up %f seconds\n', ver_str, get_uptime(aps))

stop(aps);
init(aps);

set_trigger_interval(aps, 0.02);
set_trigger_source(aps, 'internal');

fprintf('Playing square waveform\n');
wf = [zeros(100,1); ones(200,1); zeros(100,1)];
load_waveform(aps, 1, wf);
load_waveform(aps, 2, wf);
set_run_mode(aps, 'TRIG_WAVEFORM');
run(aps);
input('Press ANY key to continue: ')
stop(aps);

fprintf('Playing ramp waveforms\n');
wf = linspace(-1,1,2^14);
load_waveform(aps, 1, wf);
load_waveform(aps, 2, -wf);
run(aps);
input('Press ANY key to continue: ')
stop(aps);

fprintf('Playing cos/sin waveform\n');
load_waveform(aps, 1, ones(1200))
load_waveform(aps, 2, zeros(1200))
set_waveform_frequency(aps, 10e6)
set_run_mode(aps, 'CW_WAVEFORM');
run(aps);
input('Playing ANY key to continue: ')
stop(aps);

[example_path,~,~] = fileparts(mfilename('fullpath'));
aps.set_run_mode('RUN_SEQUENCE')

function demo_sequence(display_text, sequence_file)
    fprintf(display_text);
    load_sequence(aps, fullfile(example_path, [sequence_file, '.h5']));
    run(aps);
    input('\nPress ANY key to continue: ')
    stop(aps);
end

demo_sequence('Playing Ramsey with slipped markers', 'ramsey_slipped')

demo_sequence('Playing TPPI Ramsey sequence (modulation instructions)', 'ramsey_tppi');

demo_sequence('Playing TPPI Ramsey sequence with SSB (modulation instructions)', 'ramsey_tppi_ssb');

demo_sequence('Playing Rabi amplitude sweep (waveform prefetching)', 'rabi_amp');

demo_sequence('Playing CPMG (repeats and subroutines)', 'cpmg');

demo_sequence('Playing flip-flops (instruction prefetching)', 'instr_prefetch');

disconnect(aps);

end

File: examples/aps2_demo.py
import sys
import os
import numpy as np

#assume this is run from /examples and add ../src/python to path
import sys
import os
examples_path = os.path.abspath(os.path.dirname(__file__))
aps2_module_path = os.path.join(os.path.dirname(examples_path), "src/python")
if aps2_module_path not in sys.path:
    sys.path.append(aps2_module_path)

import aps2

aps = aps2.APS2()
aps.connect(str(sys.argv[1]))

uptime = aps.get_uptime()
firmware_version = aps.get_firmware_version()
print("Connected to APS2 running firmware version {} and up {} seconds".format(firmware_version[3], uptime))

aps.init()
aps.set_trigger_source(aps2.INTERNAL)
aps.set_trigger_interval(20e-3)

print("Playing ramp waveforms")
aps.set_waveform_float(0, np.linspace(-1,1, 2**14, dtype=np.float32))
aps.set_waveform_float(1, np.linspace(1,-1, 2**14, dtype=np.float32))
aps.set_run_mode(aps2.TRIG_WAVEFORM)
aps.run()
input("Press Enter to continue...")
aps.stop()

print("Playing modulated CW waveforms")
aps.set_waveform_float(0, np.ones(1200, dtype=np.float32))
aps.set_waveform_float(1, np.zeros(1200, dtype=np.float32))
aps.set_waveform_frequency(10e6)
aps.set_run_mode(aps2.CW_WAVEFORM)
aps.run()
input("Press Enter to continue...")
aps.stop()

aps.set_run_mode(aps2.RUN_SEQUENCE)
def demo_sequence(display_text, sequence_file):
    print(display_text)
    aps.load_sequence_file(os.path.join(examples_path, sequence_file+".h5"))
    aps.run()
    input("Press Enter to continue...")
    aps.stop()

demo_sequence("Playing Ramsey sequence with slipped markers", "ramsey_slipped")

demo_sequence("Playing Ramsey sequence with TPPI (modulation instructions)", "ramsey_tppi")

demo_sequence('Playing TPPI Ramsey sequence with SSB (modulation instructions)', 'ramsey_tppi_ssb');

demo_sequence("Playing Rabi width sweep (waveform prefetching)", "rabi_width")

demo_sequence("Playing CPMG (repeats and subroutines)", "cpmg")

demo_sequence("Playing flip-flops (instruction prefetching)", "instr_prefetch")

aps.disconnect()
